demos/dev_mode_key_provisioning/src/aws_key_provisioning_demo.c

```diff
@@ -0,0 +1,613 @@
+            {
+                configPRINTF( ( "Failed to initialize PK context with given information.\r\n" ) );
+            }
+
+            pk_cont.pk_ctx = &xEcdsaContext;
+
+
+            /* Generating CSR */
+            /* Initializing CSR Context */
+            mbedtls_x509write_csr my_csr;
+            mbedtls_x509write_csr_init( &my_csr );
+            ret = mbedtls_x509write_csr_set_subject_name( &my_csr, "CN=ThingName" );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "Failed to set subject name of CSR context.\r\n" ) );
+            }
+
+            mbedtls_x509write_csr_set_key( &my_csr, &pk_cont );
+            mbedtls_x509write_csr_set_md_alg( &my_csr, MBEDTLS_MD_SHA256 );
+            ret = mbedtls_x509write_csr_set_key_usage( &my_csr, MBEDTLS_X509_KU_DIGITAL_SIGNATURE );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "Failed to set key usage of CSR context.\r\n" ) );
+            }
+
+            ret = mbedtls_x509write_csr_set_ns_cert_type( &my_csr, MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "Failed to set NS Cert Type of CSR context.\r\n" ) );
+            }
+
+            /* Output Buffer */
+            unsigned char * final_csr = pvPortMalloc( 2000 );
+            size_t len_buf = ( size_t ) 2000;
+
+            ret = mbedtls_x509write_csr_pem( &my_csr, final_csr, len_buf, &prvRNG, &xGlobalSession );
+
+            if( ret != 0 )
+            {
+                configPRINTF( ( "Failed to write CSR.\r\n" ) );
+            }
+
+            /* Console Outputs */
+            unsigned char * csr_message = "1) PLEASE COPY THE FOLLOWING CERTIFICATE REQUEST INTO tools/create_certs/device_cert.csr :";
+            unsigned char * script_message = "2) ONCE YOU'VE COPIED THE CERTIFICATE REQUEST, PLEASE RUN THE SCRIPT NAMED\n\t\t\"device_cert.h\" LOCATED IN tools/create_certs";
+            unsigned char * cert_message = "3) ONCE YOU HAVE COMPLETED RUNNING THE SCRIPT, OPEN \"aws_clientcrediental_keys.h\" AND:"                                                                  \
+                                           "\n\ta) FORMAT THE TWO CERTIFICATES LOCATED IN tools/create_certs/deviceCertAndCACert.crt\n\t\tUSING tools/certificate_configuration/PEMfileToCString.html" \
+                                           "\n\tb) PASTE THE RESULTING TWO C STRINGS INTO\n\t\tkeyCLIENT_CERTIFICATE_PEM AND keyJITR_DEVICE_CERTIFICATE_AUTHORITY_PEM, RESPECTIVELY,\n\t\tIN \"aws_clientcrediental_keys.h\"";
+
+            configPRINTF( ( "\n\n%s\n\n%s\n\n%s\n\n%s\r\n\n", csr_message, final_csr, script_message, cert_message ) );
+
+            /* Freeing Memory */
+            mbedtls_ecp_group_free( &xEcdsaContext.grp );
+            mbedtls_ecdsa_free( &xEcdsaContext );
+            /*mbedtls_pk_free( &pk_cont ); */
+            mbedtls_x509write_csr_free( &my_csr );
+            vPortFree( final_csr );
```

Comment:
Is header_copy being leaked?

Reply:
Yes, I missed that. Added a line to free it.

Reply:
^ Still working on this. It appears that not all contents of header_copy are being freed correctly, which is causing a memory leak. Will continue to look into it. Thanks.

Reply:
I don't see the call to free header_copy. Is that in a separate commit that you haven't pushed yet?

Reply:
Yes, this has not yet been addressed. Sukhmani was working on fixing it yesterday but still needs to continue on it. There will be another patch coming up for this.